### PR TITLE
Remove need for GUI to build authentication methods

### PR DESCRIPTION
**Pull request by nirvn**
## Description

This PR removes the need for WITH_GUI to build authentication method libraries. This allows for QGIS core library users to use the authentication manager without having to build the GUI library.

### CODE REVIEW

**Comment by nirvn** [comment]
@elpaso , I'm curious, how does QGIS Server deal with authentication methods when loading projects? Obviously for oauth2 it doesn't make much sense but for e.g. basic authentication method.

**Comment by nirvn** [comment]
(unrelated OGC tests failure)

**Comment by 3nids** [comment]
good move.
shall we add a `WITH_AUTH` option?

**Comment by nirvn** [comment]
@3nids , not sure -- @nyalldawson was mentioning it earlier in a chat we had. What makes me think we don't need to is:
1/ without those methods, we might as well not have an authentication manager in src/core to begin with as it disables itself when no methods are registered;
2/ the only method that can create an additional dependency is the oauth2 method, which already has a flag to disable it (WITH_OAUTH2_PLUGIN).

So I don't see a need. But, if you and Nyall feels it's needed, very easy to add at this stage.

**Comment by nirvn** [comment]
Thinking out loud, the flag would have to be something like WITH_AUTH_METHODS or something since we always build an auth manager with core.

**Comment by nirvn** [comment]
@3nids , @nyalldawson , OK, the two of you raising the possibility of adding a WITH_AUTH was enough to convince me. Added.

**Comment by elpaso** [comment]
> @elpaso , I'm curious, how does QGIS Server deal with authentication methods when loading projects? Obviously for oauth2 it doesn't make much sense but for e.g. basic authentication method.

When loading layers it works transparently as it works on the desktop, there is some documentation about auth on the server: https://docs.qgis.org/3.16/en/docs/user_manual/auth_system/auth_workflows.html#qgis-server-support

**Comment by nirvn** [comment]
@elpaso , so the server is built with WITH_GUI=ON?

File: src/auth/esritoken/CMakeLists.txt [hunk 1]
```diff
@@ -1,38 +1,59 @@
 set(AUTH_ESRI_TOKEN_SRCS
   qgsauthesritokenmethod.cpp
-  qgsauthesritokenedit.cpp
 )
+if (WITH_GUI)
```
Comment (3nids) [inline src/auth/esritoken/CMakeLists.txt:4]: I am not a huge fan of this.
With further developments, this is most likely that someone will introduce an include of the gui in the core part.
It's a minimal effort to split this into 2 distinct targets `auth` and `auth_gui` right now and will avoid headaches (after merging PRs breaking the build of core only).

File: src/auth/esritoken/CMakeLists.txt [hunk 2]
```diff
@@ -1,38 +1,59 @@
...
+  set(AUTH_ESRI_TOKEN_HDRS ${AUTH_ESRI_TOKEN_HDRS}
+    qgsauthesritokenedit.h
+  )
+endif()
 
 include_directories (
   ../../core
   ../../core/auth
   ../../core/geometry
   ../../core/metadata
```
Comment (3nids) [inline src/auth/esritoken/CMakeLists.txt:24]: I believe that you can remove these includes now that we use target_link_librairies to qgis_core

File: src/auth/esritoken/CMakeLists.txt [hunk 3]
```diff
@@ -1,38 +1,59 @@
...
-)
-include_directories (
```
Comment (3nids) [inline src/auth/esritoken/CMakeLists.txt:19]: same for the gui

Comment (3nids) [inline src/auth/esritoken/CMakeLists.txt:19]: same for the gui

File: src/auth/esritoken/CMakeLists.txt (hunk 1 above)
Comment (nirvn) [reply src/auth/esritoken/CMakeLists.txt:4]: @3nids , I've templated this from the way we handle data providers here. Is there a good reason here to diverge from that? (I'm open minded, merely asking to get educated :) )

File: src/auth/esritoken/CMakeLists.txt (hunk 2 above)
Comment (nirvn) [reply src/auth/esritoken/CMakeLists.txt:24]: @3nids , good one.

**Pull request merged**